### PR TITLE
ttl mechanism + check/serf status filters

### PR DESCRIPTION
**Pull request by nmarasoiu**
Send heartbeats at 2/3 of ttl + List service nodes filtered by alive, when getting the server list.
When you have questions or feedback, I will continue with this about 1 hour per day,
Please check comments from prev pulls for details,
Have fun!

### CODE REVIEW

**Comment by nmarasoiu** [comment]
ttl

Sending heartbeats faster than expected ttl for 2 reasons:

    the clock drift
    consul agent should hear sooner than ttl, so that there is no race condition window in which service can be flagged as unavailable

Also, TTL is a global configuration. Setting ttl per service is not really needed in my opinion, but if you have a clean way to allow this, we can factor it in.

**Comment by nmarasoiu** [comment]
alive serf status filter

Most of the code is in a new module, that anyone can pull in maven if they want it.
No modification to existing modules is done, only a few additions of fields and remote http call methods.
I can also create an issue/feature request to track this.

The main functional idea is that a new server list filter is added in a separate module, but which needs few additions in existing core ones.

The list of server serving a service will be filtered by serf status, so that only services hosted by "alive" members of the gossip pools are listed.

File: pom.xml [hunk 1]
```diff
@@ -18,12 +18,12 @@
         <!-- lookup parent from repository -->
     </parent>
 
-	<scm>
-		<url>https://github.com/spring-cloud/spring-cloud-consul</url>
-		<connection>scm:git:git://github.com/spring-cloud/spring-cloud-consul.git</connection>
-		<developerConnection>scm:git:ssh://git@github.com/spring-cloud/spring-cloud-consul.git</developerConnection>
-		<tag>HEAD</tag>
-	</scm>
+    <scm>
```
Comment (spencergibb) [inline pom.xml]: Try not to reformat.  We have code conventions here: https://github.com/spring-cloud/spring-cloud-netflix#code-conventions-and-housekeeping-1  That will get added to this project soon.

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        });
+        Runtime.getRuntime().addShutdownHook(shutdownThread);
+        serviceIds = Collections.newSetFromMap(new ConcurrentHashMap<String, Boolean>());
+        isShuttingDown = new AtomicBoolean();
+        ttlPingThread = new TtlPingThread();
+    }
+
+    @Override
+    public void setApplicationContext(ApplicationContext context) throws BeansException {
+        ttl = context.getEnvironment().getProperty("consul.ttl", Integer.class, DEFAULT_TTL);
```
Comment (spencergibb) [inline spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java]: We should use `@ConfigurationProperties`

File: spring-cloud-consul-utils/src/main/java/org/springframework/cloud/consul/alive/AliveFilteringContext.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Injects a server list filter for giving servers hosting a service only for the live servers per serf status.
+ * @author nicu marasoiu on 10.03.2015.
+ */
+@Configuration
+@ComponentScan
```
Comment (spencergibb) [inline spring-cloud-consul-utils/src/main/java/org/springframework/cloud/consul/alive/AliveFilteringContext.java]: Libraries should use `@ComponentScan`.

File: spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    @Autowired
+    private ConsulClient client;
+
+    public TtlScheduler() {
+        if (_instances.addAndGet(1) > 1) {
+            throw new IllegalStateException("Expecting this to be used in singleton mode!");
+        }
+        //one thread to refresh ttl for each registered app to local agent is ok
+        ttlExecutor = Executors.newSingleThreadScheduledExecutor();
```
Comment (spencergibb) [inline spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java]: Spring has great facilities for handling scheduled executions: `@Scheduled`.  It handles startup, thread management and shutdown.

**Comment by nmarasoiu** [comment]
The branch did not compile even before my commits for 2 reasons, I fixed first but not the second:
- gson was missing; added gson dependency used by ecwid
- the tests need consul agent started; injecting mocked instances in the context may work

**Comment by nmarasoiu** [comment]
Summarizing, I have done all of the changes, please let me know if there is more,
And currently, after the gson fix, there are 2 problems which pre-dated my changes and seem to be related to the consul-client branch:
- the tests need consul agent started on localhost
- the following exception occurs when starting an application using the jar:
Exception in thread "main" java.lang.IllegalStateException: Could not evaluate condition on org.springframework.boot.autoconfigure.PropertyPlaceholderAutoConfiguration#propertySourcesPlaceholderConfigurer due to internal class not found. This can happen if you are @ComponentScanning a springframework package (e.g. if you put a @ComponentScan in the default package by mistake)
	at org.springframework.boot.autoconfigure.condition.SpringBootCondition.matches(SpringBootCondition.java:51)
	at org.springframework.context.annotation.ConditionEvaluator.shouldSkip(ConditionEvaluator.java:102)
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsForBeanMethod(ConfigurationClassBeanDefinitionReader.java:194)
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsForConfigurationClass(ConfigurationClassBeanDefinitionReader.java:148)
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitions(ConfigurationClassBeanDefinitionReader.java:124)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:318)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:239)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:254)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:94)
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:606)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:462)
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:118)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:686)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:320)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:957)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:946)
	at com.payu.global.wallet.service.discovery.client.Run.main(Run.java:25)
Caused by: java.lang.NoClassDefFoundError: feign/Contract
	at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2615)
	at java.lang.Class.getDeclaredMethods(Class.java:1860)
	at org.springframework.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:571)
	at org.springframework.util.ReflectionUtils.doWithMethods(ReflectionUtils.java:490)
	at org.springframework.util.ReflectionUtils.doWithMethods(ReflectionUtils.java:474)
	at org.springframework.util.ReflectionUtils.getUniqueDeclaredMethods(ReflectionUtils.java:534)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.getTypeForFactoryMethod(AbstractAutowireCapableBeanFactory.java:677)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.determineTargetType(AbstractAutowireCapableBeanFactory.java:621)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.predictBeanType(AbstractAutowireCapableBeanFactory.java:591)
	at org.springframework.beans.factory.support.AbstractBeanFactory.isFactoryBean(AbstractBeanFactory.java:1397)
	at org.springframework.beans.factory.support.AbstractBeanFactory.isFactoryBean(AbstractBeanFactory.java:968)
	at org.springframework.boot.autoconfigure.condition.BeanTypeRegistry$OptimizedBeanTypeRegistry.addBeanTypeForNonAliasDefinition(BeanTypeRegistry.java:257)
	at org.springframework.boot.autoconfigure.condition.BeanTypeRegistry$OptimizedBeanTypeRegistry.addBeanType(BeanTypeRegistry.java:246)
	at org.springframework.boot.autoconfigure.condition.BeanTypeRegistry$OptimizedBeanTypeRegistry.getNamesForType(BeanTypeRegistry.java:227)
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition.collectBeanNamesForType(OnBeanCondition.java:158)
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition.getBeanNamesForType(OnBeanCondition.java:147)
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition.getMatchingBeans(OnBeanCondition.java:119)
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition.getMatchOutcome(OnBeanCondition.java:94)
	at org.springframework.boot.autoconfigure.condition.SpringBootCondition.matches(SpringBootCondition.java:45)
	... 16 more
Caused by: java.lang.ClassNotFoundException: feign.Contract
	at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
2015-03-13 13:27:08.246  INFO 6609 --- [       Thread-1] s.c.a.AnnotationConfigApplicationContext : Closing org.springframework.context.annotation.AnnotationConfigApplicationContext@4797e7b: startup date [Fri Mar 13 13:27:04 EET 2015]; root of context hierarchy
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
	... 36 more
Disconnected from the target VM, address: '127.0.0.1:48014', transport: 'socket'

**Comment by spencergibb** [comment]
@nmarasoiu have you signed the [contributor agreeement](https://support.springsource.com/spring_committer_signup)?

**Comment by nmarasoiu** [comment]
Yes, confirmation nb 112020150313062221 altough I have put as team lead the only person I could find, namely Matt Stine
Product Manager - Spring / Cloud Foundry
Nesbit, MS

**Comment by spencergibb** [comment]
that's fine

**Pull request merged**